src/System.Linq.Parallel/tests/QueryOperators/GroupByTests.cs

```diff
@@ -0,0 +1,458 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace Test
+{
+    public class GroupByTests
+    {
+        [Theory]
+        [MemberData("Ranges", (object)(new int[] { 0, 1, 2, 7, 8, 15, 16 }), MemberType = typeof(UnorderedSources))]
+        public static void GroupBy_Unordered(Labelled<ParallelQuery<int>> labelled, int count)
+        {
+            ParallelQuery<int> query = labelled.Item;
+            IntegerRangeSet groupsSeen = new IntegerRangeSet(0, Math.Min(count, 8));
```

Comment:
As noted elsewhere, it'd be good to move constants like this 8 into a named constant.